[WPE] WPE Platform: handle multiple next presentation update callbacks

carlosgcampos opened the pull request:
#### 48d151b0fd825ae35cd7996ae211fb89d282210f
<pre>
[WPE] WPE Platform: handle multiple next presentation update callbacks
<a href="https://bugs.webkit.org/show_bug.cgi?id=292533">https://bugs.webkit.org/show_bug.cgi?id=292533</a>

Reviewed by Nikolas Zimmermann.

We currently have a limit of 1, but some layout tests end up calling
callAfterNextPresentationUpdate before the previous one is completed, so
we can just handle multiple callbacks like gtk port does.

* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
(WKWPE::ViewPlatform::~ViewPlatform):
(WKWPE::ViewPlatform::dispatchPendingNextPresentationUpdateCallbacks):
(WKWPE::ViewPlatform::callAfterNextPresentationUpdate):
* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.h:

Canonical link: <a href="https://commits.webkit.org/294553@main">https://commits.webkit.org/294553@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/51a4f1b6f4f32227d870dc74b203a0d8ce72c861

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102119 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21786 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12102 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107278 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52754 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22095 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30293 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77725 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34720 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105125 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17098 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92198 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58060 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16926 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52112 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86767 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10296 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109652 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29251 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21567 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86695 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29613 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88402 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86278 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31081 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8803 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23489 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16616 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29179 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34474 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28990 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32313 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30549 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.h:
```diff
@@ -81,14 +82,16 @@ class ViewPlatform final : public View {
     uint32_t m_displayID { 0 };
     unsigned long m_bufferRenderedID { 0 };
-    CompletionHandler<void()> m_nextPresentationUpdateCallback;
+    Vector<CompletionHandler<void()>> m_nextPresentationUpdateCallbacks;
```
I find this Vector is used in a custom way: we insert at the front and retrieve from the back, the opposite of regular queues where you queue at the back and take from the front.

I understand we do this because only `takeLast` exists (and not `takeFirst`), but find that `m_next` is confusing: `at(0)` would actually be the `last` and not the `next`...

`m_frontInsertedPresentationUpdateCallbacks`?

carlosgcampos replied:
next is because this comes from `scheduleAfterNextPresentationUpdate` so all the callbacks are called the next presentation update.

carlosgcampos committed:
[WPE] WPE Platform: handle multiple next presentation update callbacks
https://bugs.webkit.org/show_bug.cgi?id=292533

Reviewed by Nikolas Zimmermann.

We currently have a limit of 1, but some layout tests end up calling
callAfterNextPresentationUpdate before the previous one is completed, so
we can just handle multiple callbacks like gtk port does.

* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.cpp:
(WKWPE::ViewPlatform::~ViewPlatform):
(WKWPE::ViewPlatform::dispatchPendingNextPresentationUpdateCallbacks):
(WKWPE::ViewPlatform::callAfterNextPresentationUpdate):
* Source/WebKit/UIProcess/API/wpe/WPEWebViewPlatform.h:

Canonical link: https://commits.webkit.org/294553@main
